### ISSUE TITLE
Unable to reproduce Figure 2 convergence speedup

### ISSUE THREAD

**Issue by siantonelli** [Jiachen-T-Wang/GhostSuite#4]
Hi,

I’m trying to reproduce Figure 2 in Section 5.3.1 of the paper, where filtering out negative data-value samples yields ~25% faster convergence. But in my experiments, removing negative-value samples has no measurable effect on convergence.

I’m using the first-order approximation. Following Section 5.3.1, I’m training GPT-2 small for slightly more steps (50k instead of 20k) to let it reach a comparable loss. My training setup: learning rate 6e-4, batch size 16, val batch size 1, and a 2k warmup (per Appendix E).

For the values, I sum the saved dot products in the `dot_prod_log_iter_*.pt` files for the same sample ids across training steps.. To exclude negative samples, I modified the dataloader so that if a batch contains an idx in my negative-value list, it re-samples until all samples are non-negative. After training on this filtered dataset, I get the following comparison of test losses:
<img width="573" height="441" alt="Image" src="https://github.com/user-attachments/assets/d257039e-676d-474c-904f-162aa62befab" />

There are a few clear mismatches with the paper’s results:
- The test loss values are noticeably higher.
- I find ~34% negative-valued examples (over the processed data), whereas the paper reports ~16%.

Do you see any potential issue in my pipeline that might prevent reproducing Figure 2? Any pointers would be appreciated.

**Comment by Jiachen-T-Wang** [comment]
Hi! You need to use a larger validation batch size and sample the validation batch differently in each iteration when computing the ghost dot-product (otherwise, you will overfit to the specific val points you use to compute the data values). 

Also, nanoGPT's implementation for GPT2 is much better in both performance and efficiency, if I remember correctly. https://github.com/karpathy/nanoGPT/blob/master/model.py

**Comment by siantonelli** [comment]
Hi @Jiachen-T-Wang,

I've been playing around with the code to replicate Figure 2.

First, I successfully reproduced the test loss convergence for the original training run (blue line). I submitted a PR with the necessary fixes here: https://github.com/Jiachen-T-Wang/GhostSuite/pull/5#issue-3690962858.

However, I am still stuck on reproducing the ~25% faster convergence with filtering (red line).

I followed your suggestion to increase the validation batch size to 16 and sample a fresh batch at every step. I then tried three different strategies to handle the filtered batches (i.e., the removal of negative-value samples):

* Replacing filtered samples with new, positive-value samples to maintain a constant batch size. <img width="572" height="441" alt="Image" src="https://github.com/user-attachments/assets/4113ae1a-9758-448b-84ab-5164bcd41a10" />

* Simply removing the negative samples and training with the remaining smaller batch. <img width="572" height="441" alt="Image" src="https://github.com/user-attachments/assets/de5b201e-948d-4ab7-ae1b-bff4f0ac4640" />

* Filling the gaps with duplicates of existing positive samples in the batch. <img width="572" height="441" alt="Image" src="https://github.com/user-attachments/assets/be9ad964-7bca-49da-a5ec-4f70d2ff7473" />

Unfortunately, none of these methods showed the expected improvement. Could you share more details on this experiment? Like how did you handle the batch filtering? Maybe you can help me in spotting the bug :)

Thanks again!

**Comment by ZihaoWu305** [comment]
Hi, thanks for the amazing work. 

I’ve encountered a similar issue when trying to reproduce the results. I set the validation batch size to 16 and was able to match the ~16% negative samples reported in the paper. However, after following the sampling strategy suggested by @siantonelli, I did not observe any improvement in convergence speed either.

There must be some key details that we are missing. I’d really appreciate any guidance—thanks in advance!

**Comment by Jiachen-T-Wang** [comment]
Sorry for the late response, I am currently traveling (at NeurIPS last week).

If I understand correctly, it sounds like in the second and third plots you're directly comparing the original 5k-step training run against a 5k-step run that filters out negative data. This would mean the second run has fewer training tokens, which would very likely perform worse simply due to the reduced tokens.

In the experiment, we first conduct a full training run for 20,000 iterations and compute In-Run Data Shapley scores over all the data points used during training. We then filter out data points with negative contributions and train a new model on the remaining dataset for 5,000 iterations, keeping all hyperparameters the same. Importantly, the "original" curve in our plot refers to the first 5k steps of the original 20k training run—so in those first 5k steps, both models are trained on the same number of tokens. This differs from a data pruning experiment setting. In-Run Data Shapley is not designed for data selection; the purpose of this experiment is to demonstrate that pretraining corpora still have significant room for improvement in quality. 

Is this the same setup as your first plot? I also saw that in PR you set '--gradient_accumulation_steps 30', which means that your batch size is 16*30=480. Not sure how that will affect ghost dot-product computation. I also noticed while re-reading the paper that Appendix E.3 incorrectly states "10,000 iterations" when it should say "5,000"—I missed updating this during revision. Thanks for the question that prompted me to catch this!

Let me know if this clarifies things and whether you can reproduce similar results.

**Comment by siantonelli** [comment]
Thanks for the explanation.

Yes, in those plots I trained the original model for 20k steps and compared it against the filtered run (10k steps), plotting only the first 5k. I now see the issue you pointed out: by filtering dynamically at the batch level, I likely introduced indices that were never seen (and thus never scored) during the original 20k run.

If I understand the setting correctly now, for the 5k run I should sample exclusively from the pool of positive-valued samples identified in the original run. This guarantees that at step 5k both models have processed the exact same number of tokens, but the filtered run consumes only "high-quality" ones.
If that's the case, I just have one question. Does the order in which these high-quality samples are seen matter for convergence?

Regarding gradient accumulation, I included it to match the nanoGPT effective batch size and it was actually the only way I was able to reproduce the original convergence in Figure 2. The dot product computation should be safe because the results are appended at each micro_step so nothing is overwritten.

**Comment by siantonelli** [comment]
A quick update: I implemented the sampling strategy using only the positive samples as discussed to ensure the seen tokens were identical for both runs. Unfortunately, the resulting loss curve for the filtered run is still indistinguishable from the baseline.

<img width="623" height="463" alt="Image" src="https://github.com/user-attachments/assets/79b80dc0-a34c-4f0a-8a6e-3e7c87307651" />

Did you use a significantly larger validation batch size in your experiments to get a more valuable signal? Or did you perhaps ensure the same ordering of training batches between the two runs?

**Comment by Jiachen-T-Wang** [comment]
Interesting! It's surprising that the two curves are identical. I remember I tried both random and fixed ordering previously, and they didn't show much difference. Please feel free to share your implementation—I'd be happy to review it during the holiday. In the meantime, you might try tuning the training hyperparameters, particularly the learning rate. From the plot, it appears the data distributions between the original and new training runs are very similar.

**Comment by siantonelli** [comment]
Thanks for your availability! I'll play around with the hyperparameters and let you know if I can get better convergence.

In the meantime, you can find the code with the fixes in this branch of my fork: https://github.com/siantonelli/GhostSuite/tree/fix/faster-convergence

**Comment by Jiachen-T-Wang** [comment]
Hi, during the last round of code review, I identified a bug in the gradient dot-product calculation. I have pushed a cleaned version of the codebase that fixes this issue. Also, I have implemented support for training with Shapley value-filtered data points. I haven’t fully checked the rest of the code yet, so there may still be subtle bugs or mismatch with the original experiments (this is a new codebase I wrote for easier use and future development). I have performed preliminary testing on the following runs:
(1) Training run that computes In-Run Data Shapley:
```bash
cd Examples/GradDotProd_LM
./train.sh --learning_rate 0.0006 --max_steps 10000
```
(2) Training run that trains on Shapley value-filtered data points. 
```bash
./train.sh --learning_rate 0.0006 --max_steps 5000 --method Regular --replay_run_dir /path/to/last/run
```
These two runs produced the following validation loss curve:
<img width="1811" height="537" alt="Image" src="https://github.com/user-attachments/assets/e606e847-998c-4ce3-a91e-02cb1f06c6db" />
